Expose VM/VMI readiness via a dedicated CLI column

zcahana opened the pull request:
**What this PR does / why we need it**:

Add a `READY` column to the tabular output of `kubectl get vm` and `kubectl get vmi`, indicating the readiness condition of the VMI.

Here's how the output looks now:
```sh
zvic@ZVICAHANA-D2LK:~$ kubectl get vm
NAME     AGE   STATUS    READY
cirros   30s   Running   True

zvic@ZVICAHANA-D2LK:~$ kubectl get vmi
NAME     AGE   PHASE     IP            NODENAME   READY
cirros   37s   Running   10.244.0.74   minikube   True

zvic@ZVICAHANA-D2LK:~$ kubectl get vmi -o wide
NAME     AGE   PHASE     IP            NODENAME   READY   LIVE-MIGRATABLE   PAUSED
cirros   39s   Running   10.244.0.74   minikube   True    True
```

**Special notes for your reviewer**:

1. ~Note the slight casing difference between the VM and VMI outputs. The VM column feeds from the `.status.ready` field which is in lowercase, whereas the VMI column feeds from the status field of the VMI's "Ready" condition~ (no longer relevant as the VM column now too feeds from the VM's "Ready" condition).
2. I've used this chance to remove the obsolete `VOLUME` column from the output of `kubectl get vm`.
  This column, added way back in #1654, was pointing to a non-existing VM field, `vm.spec.volumes[0]`, and anyway incorrectly 
  assumed that the first listed volume for the VM is the "primary" one (although a provided boot order may indicate otherwise).
3. Also removed the "CREATED" column from `kubectl get vm`. With the recent introduction of the "STATUS" column, it has became quite redundant.

**Release note**:
```release-note
Added a READY column to the tabular output of "kubectl get vm/vmi"
```

zcahana commented:
/cc @kwiesmueller

kwiesmueller commented:
Thanks for adding this!
I wonder if a ready indicator should also be added to ReplicaSets (if we don't have them there already).
Also I'm not that excited about the casing differences you mentioned.
Does the VM not have a respective condition we can use as well?

/assign

zcahana commented:
> It doesn't. Should we add one?

My bad, it does have one. We can use it. Downside is, we'll have the READY column specifying `True`, while the `CREATED` column specifies `true` - at least until we have a corresponding Created condition. We can also consider removing the `CREATED` column, as it's quite redundant with the new `STATUS` column.

kwiesmueller commented:
Hm. Can we set a default value for the column which would be false instead of empty?
I'm fine with ditching `CREATED` in favor of this, but would like to hear other people on this.

zcahana commented:
> The column just points to the .status.conditions["Ready"].status, which only gets populated at some later stage. We could modify the VM controller to set this value to false when the VMI doesn't exist or doesn't carry the condition, and modify the VMI controller to set this value to false before the virt-launcher pod is created.

/cc @rmohr @davidvossel since you were both involved in the original implementation of the Ready condition, I'd appreciate your feedback here, and particularly about the changes to when the Ready condition is present (i.e., is there a good reason to remove this condition instead of just setting it to `status: False`?).

> Downside is, we'll have the READY column specifying True, while the CREATED column specifies true - at least until we have a corresponding Created condition. We can also consider removing the CREATED column, as it's quite redundant with the new STATUS column.

...and also regarding this matter :-)

/hold
There's additional work to be done on the VMI's unit-tests (mostly - aligning the current tests to not fail when the Ready condition is updated). I'm holding off to this work until we get an ack that these changes are acceptable.

iholder-redhat commented:
Hey @zcahana, I see this PR is still in progress.
Will review then it goes "ready to review" (stops being draft).

zcahana commented:
> I see this PR is still in progress.
Will review then it goes "ready to review" (stops being draft).

It's mostly ready to review, except for alignment of some existing unit tests that now fail due to a change in how we set the Ready condition (see [this comment](https://github.com/kubevirt/kubevirt/pull/5791#issuecomment-861478929)). I've been waiting with this in order to get some tentative ack about this change, and marked as WIP to avoid running the CI lanes in vain.
Anyway I'll find some time soon to complete these unit tests.

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -634,6 +607,65 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+		setVMICondition(k8sv1.ConditionFalse,
+			virtv1.PodTerminatingReason,
+			"virt-launcher pod is terminating",
+			v1.Now(),
```
can `v1.Now()` just get called once in this function, stored in a var, and used everywhere we need to represent `now` as a time stamp on a condition

zcahana replied:
Sure, will change accordingly.

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -634,6 +607,65 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+	if pod == nil || isTempPod(pod) {
+		setVMICondition(k8sv1.ConditionFalse,
+			virtv1.PodConditionMissingReason,
+			"virt-launcher pod does not exist",
```
What would you think about us being a bit more detailed here with the message?

for example, If the VMI is in the early phases, the message should be `virt-launcher pod has not yet been scheduled`, if we're in the later phases of a VMI the message should be something like`virt-launcher pod is terminated`

zcahana replied:
So, we already have several different messages for `Ready: False`:
- "virt-launcher pod does not exist"
- "virt-launcher pod is terminating"
- virt-launcher pod is missing the Ready condition"
- `podReadyCond.Message`, in case virt-launcher pod exists && has this condition set (that's typically ""

I've reworded the first one to "virt-launcher pod has not yet been scheduled" per your suggestion.
Do you see any other case we should have a specific message for?

davidvossel reviewed (commented):
> @davidvossel since you were both involved in the original implementation of the Ready condition, I'd appreciate your feedback here, and particularly about the changes to when the Ready condition is present (i.e., is there a good reason to remove this condition instead of just setting it to status: False?).

I think setting ready to `False` with a clear message is probably the way to go. That aligns with what I'm seeing with pod behavior. 

We probably simply deleted the ready condition in the past because it was convenient. In general, the more expressive we can make a condition, the better.

davidvossel reviewed (approved):
/lgtm
/approve

zcahana commented:
> /lgtm
/approve

Thanks @davidvossel 
I still need to spend some time on this to adjust existing unit tests that fail due a (previously) unexpected UpdateStatus with the ready condition. Will do that in the upcoming days.

dhiller commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -636,6 +527,175 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+func (c *VMIController) syncReadyConditionFromPod(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) {
+	conditionManager := controller.NewVirtualMachineInstanceConditionManager()
+
+	setVMICondition := func(status k8sv1.ConditionStatus, reason, message string, lastProbeTime, lastTransitionTime v1.Time) {
```
nit
```suggestion
	setVMIReadyCondition := func(status k8sv1.ConditionStatus, reason, message string, lastProbeTime, lastTransitionTime v1.Time) {
```

dhiller reviewed (commented):
@zcahana is this PR ready? (it looks like you did what you intended to do)

It looks like at least there's a `make generate` missing here somehow. However, otherwise besides small readability nits this looks good to me!

zcahana commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -1419,31 +1419,52 @@ func (c *VMController) isVirtualMachineStatusMigrating(vm *virtv1.VirtualMachine
-		copyConditionDetails(vmiReadyCond, vmReadyCond)
-	} else if vmReadyCond != nil && vmiReadyCond == nil {
-		log.Log.Object(vm).V(4).Info("Removing ready condition")
+	setVMCondition := func(status k8score.ConditionStatus, reason, message string, lastProbeTime, lastTransitionTime v1.Time) {
```
Changed both, thanks.

zcahana commented:
>  is this PR ready? (it looks like you did what you intended to do)

It looks like at least there's a make generate missing here somehow. However, otherwise besides small readability nits this looks good to me!

Unfortunately not yet. There are still quite some tests which are failing and needs to be aligned to this slightly revised conditions logic. Apparently, many tests (dozens of them) have/had fragile dependencies on VM/VMI conditions (e.g., they expect something on `.Conditions[0]`, or expect no conditions to exist at all) and now fail when the Ready condition is being set all the time). I'm working through these errors one-by-one, but need some more time to complete.

zcahana commented:
@davidvossel @dhiller this PR is now ready for re-review. The main changes since the previous approval are:
- Ensure we PATCH the VMI when it's handed over to virt-handler (the initial implementation used UPDATE in some cases). See bd2d87b.
- Various fixes to unit and functional tests which had fragile dependencies on `.status.conditions`.
- Some additional minor changes.

Thanks.

kwiesmueller commented on pkg/virt-controller/watch/vm.go:
```diff
@@ -1419,31 +1419,52 @@ func (c *VMController) isVirtualMachineStatusMigrating(vm *virtv1.VirtualMachine
-		copyConditionDetails(vmiReadyCond, vmReadyCond)
-	} else if vmReadyCond != nil && vmiReadyCond == nil {
-		log.Log.Object(vm).V(4).Info("Removing ready condition")
+	setVMReadyCondition := func(status k8score.ConditionStatus, reason, message string, lastProbeTime, lastTransitionTime v1.Time) {
```
could we move this in a dedicated function to make this whole method more readable?

kwiesmueller commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -652,6 +543,175 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+func (c *VMIController) syncReadyConditionFromPod(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) {
+	conditionManager := controller.NewVirtualMachineInstanceConditionManager()
+
+	setVMIReadyCondition := func(status k8sv1.ConditionStatus, reason, message string, lastProbeTime, lastTransitionTime v1.Time) {
```
could we move this in a dedicated function to make this whole method more readable?
maybe even reuse parts from above?

zcahana replied:
Done for both.

zcahana commented:
/retest

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -477,158 +480,46 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 
-				}
-			}
+	// VMI is owned by virt-controller, so patch instead of update
```
I think this should read

`VMI is owned by virt-handler`

zcahana replied:
fixed.

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -477,158 +480,46 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
-			}
-
-			_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(patch))
+		if len(patchBytes) > 0 {
+			_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(patchBytes))
```
why was this moved outside of the switch case? This is only relevant when the vmi is in the Running phase.

zcahana replied:
This was initially done since the updating of the Ready condition now occurs in multiple phases, and not only when Running as before (i.e., the `syncReadyConditionFromPod` method is called outside of the switch case block). When testing locally, it appeared to me that virt-controller/virt-handler were racing in updating the VMI (both logs showed many occurences of "_the object has been modified; please apply your changes to the latest version and try again_"). In practice I'm not convinced whether it's possible for the Ready condition to be updated when in Scheduled phase or not, but I figured it would anyway be useful to decouple from the logic that updates the local state of the object (the switch case), and the logic that determines how to update the remote object (patch vs. update).

zcahana replied:
@davidvossel I think this remains the only unresolved discussion in this PR. Please see my above comment, and let me know if you prefer me to undo that change. Thanks! :-)

davidvossel replied:
i think it's fine to keep it as you have it now

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -477,158 +480,46 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
-				}
-			}
+	// VMI is owned by virt-controller, so patch instead of update
+	if vmi.IsRunning() || vmi.IsScheduled() {
```
when the vmi is "Scheduled", virt-controller doesn't make any changes to it. This `if` statement makes it look like changes might occur when in the Scheduled phase, which isn't the case.

zcahana replied:
See prev comment.

p.s., would it read better if we do something like:
```go
if ownedByVirtHandler(vmi) {
  // prepare patch
}
```

davidvossel replied:
the more I look at this, the less I'm convinced it matters. This function is in need of refactor. your work here with putting the condition logic into a separate function is a good start

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -652,6 +543,163 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+	}
+
+	patchBuffer := bytes.Buffer{}
+	patchBuffer.WriteString("[ ")
+	for i, entry := range patchOps {
+		patchBuffer.WriteString(entry)
+		if i == len(patchOps)-1 {
+			patchBuffer.WriteString(" ]")
+		} else {
+			patchBuffer.WriteString(", ")
+		}
+	}
```
there's a function to do this for us now `GeneratePatchBytes` in the controller pkg

zcahana replied:
cool, changed to use it.

davidvossel reviewed (commented):
looks good, i just had some minor comments about the restructuring in the vmiUpdateStatus function

zcahana commented:
/retest

davidvossel reviewed (approved):
/lgtm
/approve

zcahana commented:
/unhold

The hold label was a leftover from a while back, forgot to remove it.

Thanks to all reviewers!
